### PR TITLE
Add second heuristic for calculating user did complete step

### PR DESCRIPTION
**Pull request by bsudekum**
Looking to fix: https://github.com/mapbox/mapbox-navigation-ios/issues/199

The gist of this:
* If we do not calculate the users course is with x degrees of the exit bearing for a step, we should have a fallback to another heuristic.
* Now, we check
  * if the user is not on the current step
  * if they are also on the next step
  * And have entered the maneuver radius on this step
* If so, we can assume they have now on the next step and we increment the step counter.

/cc @mapbox/navigation

### CODE REVIEW

File: MapboxCoreNavigation/RouteController.swift [hunk 1]
```diff
@@ -216,8 +263,8 @@ extension RouteController: CLLocationManagerDelegate {
             
             if routeProgress.currentLegProgress.upComingStep?.maneuverType == ManeuverType.arrive {
                 alertLevel = .arrive
-            } else if courseMatchesManeuverFinalHeading {
-                routeProgress.currentLegProgress.stepIndex += 1
+            } else if courseMatchesManeuverFinalHeading && false {
```
Comment (bsudekum) [inline MapboxCoreNavigation/RouteController.swift]: This is initentially false so we always fallback to the code that is implemented in this PR. Will revert this once we're good with the code.

File: MapboxCoreNavigation/RouteController.swift [hunk 2]
```diff
@@ -38,6 +38,8 @@ open class RouteController: NSObject {
         }
     }
     
+    var isInsideManeuverRadius = false
```
Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: It's called a maneuver zone; the zone would have a radius.

File: MapboxCoreNavigation/RouteController.swift [hunk 3]
```diff
@@ -162,7 +164,48 @@ extension RouteController: CLLocationManagerDelegate {
...
+        let isCloseToCurrentStep = newLocation.isWithin(radius, of: routeProgress.currentLegProgress.currentStep)
+        
+        // If the user is moving away from the maneuver location
+        // and they are close to the next step
+        // we can safely say they have completed the maneuver.
+        // This is intended to be a fallback case when we do find
+        // that the users course matches the exit bearing.
+        if let upComingStep = routeProgress.currentLegProgress.upComingStep {
+            let isCloseToUpComingStep = newLocation.isWithin(radius, of: upComingStep)
+            if !isCloseToCurrentStep && isCloseToUpComingStep && isInsideManeuverRadius {
```
Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: Why does this behavior depend on being within the maneuver zone? If somehow the user has made it to the next step without a location update coming in as they pass through the maneuver zone, we probably would want to advance to the next step regardless of the maneuver zone. You could imagine this case becoming relevant for roundabouts and large intersections.

Comment (bsudekum) [inline MapboxCoreNavigation/RouteController.swift]: @1ec5 I'm current guarding against the case where the current step crosses over the following step.

Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: That's great, but as I mentioned above, what if the device is unable to update its location while within the maneuver zone but the user has nonetheless proceeded to the next step? (Maybe they passed through the intersection quickly.) What if the intersection happens to be larger than the maneuver zone? (We should probably have different maneuver zone radii for different types of maneuvers.)

Comment (bsudekum) [inline MapboxCoreNavigation/RouteController.swift]: @1ec5 i think there are tradeoffs for both methods here.

**Pull request merged**